<a href="https://colab.research.google.com/github/RamzNN/RamzNN/blob/main/filter_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
df = pd.read_excel('/content/drive/MyDrive/Colab Notebooks/–†–∞–±–æ—Ç–∞/–ü–ª–∞—Ç—Ñ–æ—Ä–º–∞-–ø–∞—Ä—Å–µ—Ä "–°–∞–ø—Ñ–∏—Ä"/stage_one/–≤—ã–≥—Ä—É–∑–∫–∞_–≤–∫.xlsx')

In [3]:
df.tail()

group_id                                            name description  \
54278  56230843  ?? –§–ö –°–ø–∞—Ä—Ç–∞–∫ –ú–æ—Å–∫–≤–∞ ?? –ö—Ä–∞—Å–Ω–æ-–±–µ–ª—ã–π –ù–ê–í–°–ï–ì–î–ê!         NaN   
54279  62454446                                   –†–æ–ª–µ–≤–∞—è —Ñ—É—Ä—Ä–∏         NaN   
54280  60343205                                   –ù–û–í–ò–ô –ü–û–†–Ø–î–û–ö         NaN   
54281  56146201                   –í–µ–ª–∏–∫—ñ –í—É–π–∫–∏/–û—Å—Ç–∞–Ω–Ω—ñ–π –º–æ—Å–∫–∞–ª—å         NaN   
54282  55284725                              –ö–æ–º–∞–Ω–¥–∞ –ù–∞–≤–∞–ª—å–Ω–æ–≥–æ         NaN   

      Subscribers Count  relevance Threat assesment status    city  country  \
54278              4699       True             destructive  –ú–æ—Å–∫–≤–∞   –†–æ—Å—Å–∏—è   
54279              6024       True             destructive     NaN   –†–æ—Å—Å–∏—è   
54280             11818       True             destructive    –ö–∏–µ–≤  –£–∫—Ä–∞–∏–Ω–∞   
54281             23295       True             destructive    –ö–∏–µ–≤  –£–∫—Ä–∞–∏–Ω–∞   
54282            205372       True             destructive     NaN      NaN   

                      tags                  subcultures  
54278           –§–ö –°–ø–∞—Ä—Ç–∞–∫  –ì—Ä—É–ø–ø—ã —Ñ—É—Ç–±–æ–ª—å–Ω—ã—Ö —Ö—É–ª–∏–≥–∞–Ω–æ–≤  
54279                –§—É—Ä—Ä–∏        –ê–Ω—Ç–∏—Å–µ–º–µ–π–Ω—ã–µ —Ü–µ–Ω–Ω–æ—Å—Ç–∏  
54280  –£–∫—Ä–∞—ó–Ω—Å—å–∫–∏–π –ø–æ—Ä—è–¥–æ–∫      –ü–æ–ª–∏—Ç–∏—á–µ—Å–∫–∏–π —ç–∫—Å—Ç—Ä–µ–º–∏–∑–º  
54281              –ú–æ—Å–∫–∞–ª—å      –ü–æ–ª–∏—Ç–∏—á–µ—Å–∫–∏–π —ç–∫—Å—Ç—Ä–µ–º–∏–∑–º  
54282            –ù–∞–≤–∞–ª—å–Ω—ã–π      –ü–æ–ª–∏—Ç–∏—á–µ—Å–∫–∏–π —ç–∫—Å—Ç—Ä–µ–º–∏–∑–º

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54283 entries, 0 to 54282
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   group_id                 53583 non-null  object
 1   name                     54280 non-null  object
 2   description              40436 non-null  object
 3   Subscribers Count        53887 non-null  object
 4   relevance                54283 non-null  bool  
 5   Threat assesment status  53968 non-null  object
 6   city                     51682 non-null  object
 7   country                  52236 non-null  object
 8   tags                     54283 non-null  object
 9   subcultures              54283 non-null  object
dtypes: bool(1), object(9)
memory usage: 3.8+ MB


In [11]:
# df.drop_duplicates(subset=['name', 'description', 'relevance'], inplace=True)
# df.info()

In [5]:
# –ü—Ä–æ–≤–µ—Ä–∫–∞ –ø—Ä–æ–ø—É—Å–∫–æ–≤
print("\n–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –ø—Ä–æ–ø—É—Å–∫–æ–≤ –ø–æ —Å—Ç–æ–ª–±—Ü–∞–º:")
print(df.isnull().sum())

# –†–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ —Ü–µ–ª–µ–≤–æ–π –ø–µ—Ä–µ–º–µ–Ω–Ω–æ–π 'relevance'
print("\n–†–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ 'relevance':")
print(df['relevance'].value_counts())


–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –ø—Ä–æ–ø—É—Å–∫–æ–≤ –ø–æ —Å—Ç–æ–ª–±—Ü–∞–º:
group_id                     700
name                           3
description                13847
Subscribers Count            396
relevance                      0
Threat assesment status      315
city                        2601
country                     2047
tags                           0
subcultures                    0
dtype: int64

–†–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ 'relevance':
relevance
False    45142
True      9141
Name: count, dtype: int64


In [6]:
# Если name пустое — заполним пустой строкой
df['name'] = df['name'].fillna('')
# Если description пустое — заполним пустой строкой
df['description'] = df['description'].fillna('')

In [7]:
# –ü—Ä–æ–≤–µ—Ä–∫–∞ –ø—Ä–æ–ø—É—Å–∫–æ–≤
print("\n–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –ø—Ä–æ–ø—É—Å–∫–æ–≤ –ø–æ —Å—Ç–æ–ª–±—Ü–∞–º:")
print(df.isnull().sum())

# –†–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ —Ü–µ–ª–µ–≤–æ–π –ø–µ—Ä–µ–º–µ–Ω–Ω–æ–π 'relevance'
print("\n–†–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ 'relevance':")
print(df['relevance'].value_counts())


–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –ø—Ä–æ–ø—É—Å–∫–æ–≤ –ø–æ —Å—Ç–æ–ª–±—Ü–∞–º:
group_id                    700
name                          0
description                   0
Subscribers Count           396
relevance                     0
Threat assesment status     315
city                       2601
country                    2047
tags                          0
subcultures                   0
dtype: int64

–†–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ 'relevance':
relevance
False    45142
True      9141
Name: count, dtype: int64


In [8]:
# –û–±—ä–µ–¥–∏–Ω–µ–Ω–∏–µ –Ω–∞–∑–≤–∞–Ω–∏—è –∏ –æ–ø–∏—Å–∞–Ω–∏—è
df['text'] = df['name'].astype(str) + ' ' + df['description'].astype(str)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54283 entries, 0 to 54282
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   group_id                 53583 non-null  object
 1   name                     54283 non-null  object
 2   description              54283 non-null  object
 3   Subscribers Count        53887 non-null  object
 4   relevance                54283 non-null  bool  
 5   Threat assesment status  53968 non-null  object
 6   city                     51682 non-null  object
 7   country                  52236 non-null  object
 8   tags                     54283 non-null  object
 9   subcultures              54283 non-null  object
 10  text                     54283 non-null  object
dtypes: bool(1), object(10)
memory usage: 4.2+ MB


### –û–±—Ä–∞–±–æ—Ç–∫–∞ —Ç–µ–∫—Å—Ç–∞

–í —ç—Ç–æ—Ç —Ä–∞–∑ —è –ø–æ–ø—Ä–æ–±—É—é –±–µ–∑ –æ—á–∏—Å—Ç–∫–∏ –∏ –æ–±—Ä–∞–±–æ—Ç–∫–∏ —Ç–µ–∫—Å—Ç–∞ –Ω–æ –ø–æ—Ç–æ–º –Ω–∞–¥–æ –±—É–¥–µ—Ç —ç—Ç–æ —Å–¥–µ–ª–∞—Ç—å –∏ —Å—Ä–∞–≤–Ω–∏—Ç—å –∫–∞—á–µ—Å—Ç–≤–æ

```Python
def clean_text(text):
    # –ü—Ä–∏–≤–æ–¥–∏–º –∫ –Ω–∏–∂–Ω–µ–º—É —Ä–µ–≥–∏—Å—Ç—Ä—É
    text = text.lower()
    # –£–¥–∞–ª—è–µ–º –Ω–µ–∂–µ–ª–∞—Ç–µ–ª—å–Ω—ã–µ —Å–∏–º–≤–æ–ª—ã (–º–æ–∂–Ω–æ –∫–∞—Å—Ç–æ–º–∏–∑–∏—Ä–æ–≤–∞—Ç—å)
    text = re.sub(r'[^а-яa-z0-9ё\s.,!?–-]', ' ', text)
    # –£–±–∏—Ä–∞–µ–º –ª–∏—à–Ω–∏–µ –ø—Ä–æ–±–µ–ª—ã
    text = re.sub(r'\s+', ' ', text).strip()
    return text

df['text'] = df['text'].apply(clean_text)

```

In [10]:
print(df['relevance'].value_counts())

relevance
False    45142
True      9141
Name: count, dtype: int64


In [11]:
df['text_length'] = df['text'].apply(lambda x: len(x.split()))
df['text_length'].describe()

count    54283.000000
mean        65.324061
std        112.336866
min          1.000000
25%          4.000000
50%         18.000000
75%         71.000000
max        734.000000
Name: text_length, dtype: float64

In [12]:
from sklearn.model_selection import train_test_split

train_df, test_df = train_test_split(df, test_size=0.1, random_state=42, stratify=df['relevance'])
train_df, val_df = train_test_split(train_df, test_size=0.1, random_state=42, stratify=train_df['relevance'])

### –£—Å—Ç–∞–Ω–æ–≤–∫–∞ –Ω—É–∂–Ω—ã—Ö –±–∏–±–ª–∏–æ—Ç–µ–∫

In [13]:
!pip install --upgrade pip
!pip install transformers accelerate evaluate torch sentencepiece

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 30.5 MB/s eta 0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 24.1.2
    Uninstalling pip-24.1.2:
      Successfully uninstalled pip-24.1.2
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 20.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 169.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 180.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 44.9 MB/s eta 0:00:00
   ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚

In [14]:
import pandas as pd
import numpy as np
import torch

from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer
import evaluate

# –£–±–µ–¥–∏–º—Å—è, —á—Ç–æ —É –Ω–∞—Å –µ—Å—Ç—å CUDA (–µ—Å–ª–∏ –¥–æ—Å—Ç—É–ø–Ω–æ)
device = "cuda" if torch.cuda.is_available() else "cpu"
print("Using device:", device)

# –ü—Ä–µ–¥–ø–æ–ª–æ–∂–∏–º, –≤—ã —É–∂–µ –ø—Ä–æ—á–∏—Ç–∞–ª–∏ df, train_df, val_df, test_df
# –∫–∞–∫ –º—ã –¥–µ–ª–∞–ª–∏ —Ä–∞–Ω–µ–µ

Using device: cuda


In [15]:
train_df['relevance'] = train_df['relevance'].astype(int)
val_df['relevance'] = val_df['relevance'].astype(int)
test_df['relevance'] = test_df['relevance'].astype(int)

In [17]:
model_name = "ai-forever/ruBert-base"
tokenizer = AutoTokenizer.from_pretrained(model_name)

model = AutoModelForSequenceClassification.from_pretrained(
    model_name,
    num_labels=2  # –±–∏–Ω–∞—Ä–Ω–∞—è –∫–ª–∞—Å—Å–∏—Ñ–∏–∫–∞—Ü–∏—è
)
model.to(device)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at ai-forever/ruBert-base and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


BertForSequenceClassification(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(120138, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1

### –ü–æ–¥–≥–æ—Ç–æ–≤–∫–∞ –¥–∞–Ω–Ω—ã—Ö –¥–ª—è Trainer

In [25]:
MAX_LEN = 256  # –¢–∞–∫ –∫–∞–∫ —Å—Ä–µ–¥–Ω—è—è –¥–ª–∏–Ω–∞ ~65 —Å–ª–æ–≤, –Ω–æ max ~734, –≤—ã–±–µ—Ä–µ–º 512

def tokenize_function(examples):
    # examples – это dict с ключами "text" и "relevance"
    return tokenizer(
        examples["text"],
        truncation=True,
        max_length=MAX_LEN,
        padding="max_length",  # –∏–ª–∏ "longest", –Ω–æ max_length=512 –±–µ–∑–æ–ø–∞—Å–Ω–µ–µ –¥–ª—è batched training
    )


In [26]:
# –î–ª—è Trainer –Ω—É–∂–Ω–æ, —á—Ç–æ–±—ã –≤ —Å–ª–æ–≤–∞—Ä–µ –¥–ª—è –∫–∞–∂–¥–æ–π —Å—Ç—Ä–æ—á–∫–∏ –±—ã–ª–∏ —Ö–æ—Ç—è –±—ã –ø–æ–ª—è: 'text' –∏ 'labels' (–∏–ª–∏ 'label').

train_data = {
    "text": train_df["text"].tolist(),
    "labels": train_df["relevance"].tolist()
}

val_data = {
    "text": val_df["text"].tolist(),
    "labels": val_df["relevance"].tolist()
}

test_data = {
    "text": test_df["text"].tolist(),
    "labels": test_df["relevance"].tolist()
}


In [27]:
from datasets import Dataset

train_dataset = Dataset.from_dict(train_data)
val_dataset = Dataset.from_dict(val_data)
test_dataset = Dataset.from_dict(test_data)

# –ü—Ä–∏–º–µ–Ω—è–µ–º —Ç–æ–∫–µ–Ω–∏–∑–∞—Ü–∏—é
train_dataset = train_dataset.map(tokenize_function, batched=True)
val_dataset = val_dataset.map(tokenize_function, batched=True)
test_dataset = test_dataset.map(tokenize_function, batched=True)

# –§–∏–ª—å—Ç—Ä—É–µ–º –Ω–µ–Ω—É–∂–Ω—ã–µ –∫–æ–ª–æ–Ω–∫–∏ –∏ –ø–µ—Ä–µ–∏–º–µ–Ω–æ–≤—ã–≤–∞–µ–º labels -> label (–∏–ª–∏ –Ω–∞–æ–±–æ—Ä–æ—Ç)
train_dataset = train_dataset.remove_columns(["text"])
val_dataset = val_dataset.remove_columns(["text"])
test_dataset = test_dataset.remove_columns(["text"])
# –¢–µ–ø–µ—Ä—å –≤ –∫–∞–∂–¥–æ–º –¥–∞—Ç–∞—Å–µ—Ç–µ –µ—Å—Ç—å "input_ids", "attention_mask", (–≤–æ–∑–º–æ–∂–Ω–æ "token_type_ids") –∏ "labels".


Map:   0%|          | 0/43968 [00:00<?, ? examples/s]

Map:   0%|          | 0/4886 [00:00<?, ? examples/s]

Map:   0%|          | 0/5429 [00:00<?, ? examples/s]

### –ù–∞—Å—Ç—Ä–æ–π–∫–∞ Trainer

In [28]:
accuracy_metric = evaluate.load("accuracy")
f1_metric = evaluate.load("f1")

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)

    acc = accuracy_metric.compute(predictions=predictions, references=labels)
    f1 = f1_metric.compute(predictions=predictions, references=labels, average="binary")

    return {
        "accuracy": acc["accuracy"],
        "f1": f1["f1"]
    }

In [29]:
batch_size = 8  # –ú–æ–∂–Ω–æ —É–≤–µ–ª–∏—á–∏—Ç—å, –µ—Å–ª–∏ —Ö–≤–∞—Ç–∞–µ—Ç GPU (16/32).
num_epochs = 3  # –ß–∞—â–µ –≤—Å–µ–≥–æ 3-5 —ç–ø–æ—Ö —Ö–≤–∞—Ç–∞–µ—Ç –¥–ª—è fine-tuning
learning_rate = 1e-5

training_args = TrainingArguments(
    output_dir="my_sbert_model",
    evaluation_strategy="epoch",
    save_strategy="epoch",
    learning_rate=learning_rate,
    per_device_train_batch_size=batch_size,
    per_device_eval_batch_size=batch_size,
    num_train_epochs=num_epochs,
    weight_decay=0.01,
    load_best_model_at_end=True,
    metric_for_best_model="f1",  # –°–ª–µ–¥–∏–º –∑–∞ f1
    greater_is_better=True,
    save_total_limit=2,
    logging_steps=100,               # –õ–æ–≥–∏—Ä–æ–≤–∞—Ç—å –∫–∞–∂–¥—ã–µ 100 —à–∞–≥–æ–≤
    logging_dir="./logs",            # –î–∏—Ä–µ–∫—Ç–æ—Ä–∏—è –¥–ª—è –ª–æ–≥–æ–≤ (–µ—Å–ª–∏ –Ω—É–∂–Ω–∞)
    report_to="none",                # –ß—Ç–æ–±—ã –ª–æ–≥–∏ —à–ª–∏ –ø—Ä–æ—Å—Ç–æ –≤ –∫–æ–Ω—Å–æ–ª—å
    disable_tqdm=False               # –†–∞–∑—Ä–µ—à–∏—Ç—å tqdm-–ø—Ä–æ–≥—Ä–µ—Å—Å‚Äê–±–∞—Ä
    # FP16 ?
    # fp16=True,  # –ï—Å–ª–∏ GPU –ø–æ–¥–¥–µ—Ä–∂–∏–≤–∞–µ—Ç half precision, –º–æ–∂–Ω–æ —É—Å–∫–æ—Ä–∏—Ç—å
)

/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1594: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


In [31]:
from transformers import EarlyStoppingCallback

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    tokenizer=tokenizer,  # –ß—Ç–æ–±—ã Trainer –º–æ–≥ –¥–µ–ª–∞—Ç—å pad/collate
    compute_metrics=compute_metrics,
    callbacks=[EarlyStoppingCallback(early_stopping_patience=1)]
)

<ipython-input-31-e6fa9a6f883a>:3: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


In [32]:
trainer.train()

TrainOutput(global_step=16488, training_loss=0.3081199097899428, metrics={'train_runtime': 7026.2683, 'train_samples_per_second': 18.773, 'train_steps_per_second': 2.347, 'total_flos': 1.735270032310272e+16, 'train_loss': 0.3081199097899428, 'epoch': 3.0})

In [33]:
trainer.save_model("my_sbert_model/final")
tokenizer.save_pretrained("my_sbert_model/final")

('my_sbert_model/final/tokenizer_config.json',
 'my_sbert_model/final/special_tokens_map.json',
 'my_sbert_model/final/vocab.txt',
 'my_sbert_model/final/added_tokens.json',
 'my_sbert_model/final/tokenizer.json')

In [34]:
model.save_pretrained('/content/drive/MyDrive/Colab Notebooks/–†–∞–±–æ—Ç–∞/–ü–ª–∞—Ç—Ñ–æ—Ä–º–∞-–ø–∞—Ä—Å–µ—Ä "–°–∞–ø—Ñ–∏—Ä"/stage_one/my_sbert_model/final')
tokenizer.save_pretrained('/content/drive/MyDrive/Colab Notebooks/–†–∞–±–æ—Ç–∞/–ü–ª–∞—Ç—Ñ–æ—Ä–º–∞-–ø–∞—Ä—Å–µ—Ä "–°–∞–ø—Ñ–∏—Ä"/stage_one/my_sbert_model/final')


('/content/drive/MyDrive/Colab Notebooks/–†–∞–±–æ—Ç–∞/–ü–ª–∞—Ç—Ñ–æ—Ä–º–∞-–ø–∞—Ä—Å–µ—Ä "–°–∞–ø—Ñ–∏—Ä"/stage_one/my_sbert_model/final/tokenizer_config.json',
 '/content/drive/MyDrive/Colab Notebooks/–†–∞–±–æ—Ç–∞/–ü–ª–∞—Ç—Ñ–æ—Ä–º–∞-–ø–∞—Ä—Å–µ—Ä "–°–∞–ø—Ñ–∏—Ä"/stage_one/my_sbert_model/final/special_tokens_map.json',
 '/content/drive/MyDrive/Colab Notebooks/–†–∞–±–æ—Ç–∞/–ü–ª–∞—Ç—Ñ–æ—Ä–º–∞-–ø–∞—Ä—Å–µ—Ä "–°–∞–ø—Ñ–∏—Ä"/stage_one/my_sbert_model/final/vocab.txt',
 '/content/drive/MyDrive/Colab Notebooks/–†–∞–±–æ—Ç–∞/–ü–ª–∞—Ç—Ñ–æ—Ä–º–∞-–ø–∞—Ä—Å–µ—Ä "–°–∞–ø—Ñ–∏—Ä"/stage_one/my_sbert_model/final/added_tokens.json',
 '/content/drive/MyDrive/Colab Notebooks/–†–∞–±–æ—Ç–∞/–ü–ª–∞—Ç—Ñ–æ—Ä–º–∞-–ø–∞—Ä—Å–µ—Ä "–°–∞–ø—Ñ–∏—Ä"/stage_one/my_sbert_model/final/tokenizer.json')

### –û—Ü–µ–Ω–∫–∞ –Ω–∞ —Ç–µ—Å—Ç–µ

In [35]:
test_metrics = trainer.evaluate(test_dataset)
print(test_metrics)

{'eval_loss': 0.38629522919654846, 'eval_accuracy': 0.8810093939952109, 'eval_f1': 0.6075334143377886, 'eval_runtime': 82.6311, 'eval_samples_per_second': 65.702, 'eval_steps_per_second': 8.217, 'epoch': 3.0}


In [36]:
predictions = trainer.predict(test_dataset)
preds = np.argmax(predictions.predictions, axis=1)
labels = predictions.label_ids

# –ü–æ–¥—Å—á–µ—Ç –º–µ—Ç—Ä–∏–∫ –≤—Ä—É—á–Ω—É—é, –µ—Å–ª–∏ –Ω—É–∂–Ω–æ:
acc = accuracy_metric.compute(predictions=preds, references=labels)
f1 = f1_metric.compute(predictions=preds, references=labels)
print("Accuracy:", acc["accuracy"])
print("F1:", f1["f1"])


Accuracy: 0.8810093939952109
F1: 0.6075334143377886


### –ü–æ–¥–±–æ—Ä –ø–æ—Ä–æ–≥–∞ —Ñ–∏–ª—å—Ç—Ä–∞—Ü–∏–∏

#### –ü–æ–ª—É—á–∞–µ–º –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ –Ω–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–æ–Ω–Ω–æ–º (–∏–ª–∏ —Ç–µ—Å—Ç–æ–≤–æ–º) –Ω–∞–±–æ—Ä–µ

In [37]:
import numpy as np
from sklearn.metrics import recall_score

# Предположим, val_dataset (или test_dataset) — ваши данные для подбора порога
val_result = trainer.predict(val_dataset)
logits = val_result.predictions  # shape (num_samples, 2) –¥–ª—è –±–∏–Ω–∞—Ä–Ω–æ–π –∫–ª–∞—Å—Å–∏—Ñ–∏–∫–∞—Ü–∏–∏
labels = val_result.label_ids    # –∏—Å—Ç–∏–Ω–∞ (0 –∏–ª–∏ 1)

# –ü—Ä–µ–≤—Ä–∞—â–∞–µ–º –ª–æ–≥–∏—Ç—ã –≤ –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ –¥–ª—è –∫–ª–∞—Å—Å–∞ "True" (–∏–Ω–¥–µ–∫—Å 1)
import torch
probs = torch.nn.functional.softmax(torch.tensor(logits), dim=-1).numpy()
probs_true = probs[:, 1]  # –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç—å –ø—Ä–∏–Ω–∞–¥–ª–µ–∂–Ω–æ—Å—Ç–∏ –∫ –∫–ª–∞—Å—Å—É True


In [49]:
thresholds = np.linspace(0, 1, 101)  # 0.00, 0.01, 0.02, ..., 1.00

best_threshold = None
best_diff = float('inf')  # —á—Ç–æ–±—ã –º–∏–Ω–∏–º–∏–∑–∏—Ä–æ–≤–∞—Ç—å |recall_false - 0.40|
best_recall_false = None
best_recall_true = None

for thr in thresholds:
    # –ë–∏–Ω–∞—Ä–Ω—ã–µ –ø—Ä–µ–¥—Å–∫–∞–∑–∞–Ω–∏—è –ø–æ –≤—ã–±—Ä–∞–Ω–Ω–æ–º—É –ø–æ—Ä–æ–≥—É
    pred = (probs_true >= thr).astype(int)  # 1, –µ—Å–ª–∏ prob >= thr

    # –°—á–∏—Ç–∞–µ–º recall –¥–ª—è –∫–∞–∂–¥–æ–≥–æ –∫–ª–∞—Å—Å–∞
    recs = recall_score(labels, pred, average=None)
    recall_false = recs[0]  # –¥–ª—è –∫–ª–∞—Å—Å–∞ 0
    recall_true  = recs[1]  # –¥–ª—è –∫–ª–∞—Å—Å–∞ 1

    # –ü—Ä–æ–≤–µ—Ä—è–µ–º —É—Å–ª–æ–≤–∏–µ recall_true >= 0.97
    if recall_true >= 0.90:
        # –°–º–æ—Ç—Ä–∏–º, –Ω–∞—Å–∫–æ–ª—å–∫–æ recall_false –±–ª–∏–∑–æ–∫ –∫ 0.40
        diff = abs(recall_false - 0.50)
        if diff < best_diff:
            best_diff = diff
            best_threshold = thr
            best_recall_false = recall_false
            best_recall_true = recall_true

# –ü—Ä–æ–≤–µ—Ä—è–µ–º, –Ω–∞—à–ª–∏ –ª–∏ –º—ã –ø–æ–¥—Ö–æ–¥—è—â–∏–π –ø–æ—Ä–æ–≥
if best_threshold is not None:
    print(f"–ü–æ–¥—Ö–æ–¥—è—â–∏–π –ø–æ—Ä–æ–≥: {best_threshold:.2f}")
    print(f"Recall False = {best_recall_false:.3f} (—Ö–æ—Ç–µ–ª–∏ –æ–∫–æ–ª–æ 0.40)")
    print(f"Recall True  = {best_recall_true:.3f} (>= 0.97)")
    print(f"–†–∞–∑–Ω–∏—Ü–∞ –æ—Ç 0.40 –ø–æ False: {best_diff:.3f}")
else:
    print("–ù–µ —É–¥–∞–ª–æ—Å—å –Ω–∞–π—Ç–∏ –ø–æ—Ä–æ–≥, –ø—Ä–∏ –∫–æ—Ç–æ—Ä–æ–º recall_true >= 0.97 –∏ recall_false ~ 0.40.")


–ü–æ–¥—Ö–æ–¥—è—â–∏–π –ø–æ—Ä–æ–≥: 0.01
Recall False = 0.579 (—Ö–æ—Ç–µ–ª–∏ –æ–∫–æ–ª–æ 0.40)
Recall True  = 0.902 (>= 0.97)
–†–∞–∑–Ω–∏—Ü–∞ –æ—Ç 0.40 –ø–æ False: 0.079


### –≠–∫—Å–ø–µ—Ä–∏–º–µ–Ω—Ç

–í–∞—Ä–∏–∞–Ω—Ç 3: –ü–æ–¥–±–∏—Ä–∞—Ç—å –ø–æ—Ä–æ–≥ –∏–Ω–∞—á–µ (–¥–≤—É—Ö—à–∞–≥–æ–≤—ã–π –ø–æ–¥—Ö–æ–¥)

Иногда бизнес-логика звучит не «хотим recallfalse=0.4recallfalse​=0.4», а «хотим убрать 40% объектов из всего датасета, которые модель считает наиболее «подозрительными» на False». Тогда мы можем:

    –û—Ç—Å–æ—Ä—Ç–∏—Ä–æ–≤–∞—Ç—å –≤—Å–µ –æ–±—ä–µ–∫—Ç—ã –ø–æ –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ –∫–ª–∞—Å—Å–∞ True (–ø–æ –≤–æ–∑—Ä–∞—Å—Ç–∞–Ω–∏—é).
    –í–∑—è—Ç—å –Ω–∏–∂–Ω–∏–µ 40% –æ–±—ä–µ–∫—Ç–æ–≤ (—Ç.–µ. —Å –Ω–∞–∏–º–µ–Ω—å—à–µ–π –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç—å—é True) –∏ –æ–±—ä—è–≤–∏—Ç—å –∏—Ö False. –û—Å—Ç–∞–ª—å–Ω—ã—Ö (60%) –æ–±—ä—è–≤–∏—Ç—å True.
    –ü–æ—Å—á–∏—Ç–∞—Ç—å, –∫–∞–∫–æ–π recalltruerecalltrue‚Äã –ø–æ–ª—É—á–∏—Ç—Å—è. –ï—Å–ª–∏ –æ–Ω < 0.97, –∑–Ω–∞—á–∏—Ç –º–æ–¥–µ–ª—å –Ω–µ —É–º–µ–µ—Ç —Ç–∞–∫ —Ö–æ—Ä–æ—à–æ —Ä–∞–∑–ª–∏—á–∞—Ç—å.

In [51]:
import numpy as np
import torch
from sklearn.metrics import recall_score

# 1. –ü–æ–ª—É—á–∞–µ–º –ª–æ–≥–∏—Ç—ã –∏ –º–µ—Ç–∫–∏ –Ω–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–æ–Ω–Ω–æ–π / —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ:
val_result = trainer.predict(val_dataset)
logits = val_result.predictions         # shape: (num_samples, 2)
labels = val_result.label_ids           # –∏—Å—Ç–∏–Ω–Ω—ã–µ –∫–ª–∞—Å—Å—ã 0 –∏–ª–∏ 1

# 2. –ü—Ä–µ–æ–±—Ä–∞–∑—É–µ–º –ª–æ–≥–∏—Ç—ã –≤ –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ –ø—Ä–∏–Ω–∞–¥–ª–µ–∂–Ω–æ—Å—Ç–∏ –∫ –∫–ª–∞—Å—Å—É "True" (–∏–Ω–¥–µ–∫—Å=1):
probs = torch.nn.functional.softmax(torch.tensor(logits), dim=-1).numpy()
probs_true = probs[:, 1]  # –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç—å –∫–ª–∞—Å—Å–∞ True

# 3. –°–æ—Ä—Ç–∏—Ä—É–µ–º –∏–Ω–¥–µ–∫—Å—ã –≤—ã–±–æ—Ä–∫–∏ –ø–æ –≤–æ–∑—Ä–∞—Å—Ç–∞–Ω–∏—é –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ True (—Å–∞–º—ã–µ "False" - –ø–µ—Ä–≤—ã–µ):
idx_sorted = np.argsort(probs_true)
n = len(probs_true)

# 4. –ù–∞—Ö–æ–¥–∏–º, —Å–∫–æ–ª—å–∫–æ –æ–±—ä–µ–∫—Ç–æ–≤ = 40% –æ—Ç –≤—Å–µ–≥–æ –¥–∞—Ç–∞—Å–µ—Ç–∞:
cutoff = int(0.35 * n)  # –Ω–∏–∂–Ω–∏–µ 40%

# 5. –§–æ—Ä–º–∏—Ä—É–µ–º –∫–∞—Å—Ç–æ–º–Ω—ã–µ –ø—Ä–µ–¥—Å–∫–∞–∑–∞–Ω–∏—è:
#    - —Ç–µ, —É –∫–æ–≥–æ prob True –º–∏–Ω–∏–º–∞–ª—å–Ω–∞ (–ø–æ–ø–∞–¥–∞—é—Ç –≤ –Ω–∏–∂–Ω–∏–µ 40%) -> —Å—Ç–∞–≤–∏–º –∫–ª–∞—Å—Å 0 (False)
#    - –æ—Å—Ç–∞–ª—å–Ω—ã–µ 60% -> –∫–ª–∞—Å—Å 1 (True)
preds_custom = np.zeros(n, dtype=int)   # –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é 0 (False)
top_idx = idx_sorted[cutoff:]           # –∏–Ω–¥–µ–∫—Å—ã, –∫–æ—Ç–æ—Ä—ã–µ –∏–¥—É—Ç –≤ –≤–µ—Ä—Ö–Ω–∏–µ 60%
preds_custom[top_idx] = 1

# 6. –°—á–∏—Ç–∞–µ–º Recall –ø–æ –æ—Ç–¥–µ–ª—å–Ω—ã–º –∫–ª–∞—Å—Å–∞–º
#    average=None -> –≤–µ—Ä–Ω–µ—Ç [recall_for_class_0, recall_for_class_1]
rec_each = recall_score(labels, preds_custom, average=None)
recall_false = rec_each[0]
recall_true = rec_each[1]

print(f"–í—Å–µ–≥–æ –æ–±—ä–µ–∫—Ç–æ–≤: {n}")
print(f"–£–±—Ä–∞–ª–∏ (–ø—Ä–µ–¥—Å–∫–∞–∑–∞–ª–∏ False) –≤ –∏—Ç–æ–≥–µ: {cutoff} ({cutoff/n:.2%} –æ—Ç –¥–∞—Ç–∞—Å–µ—Ç–∞)")
print(f"Recall(False) = {recall_false:.3f}")
print(f"Recall(True)  = {recall_true:.3f}")


–í—Å–µ–≥–æ –æ–±—ä–µ–∫—Ç–æ–≤: 4886
–£–±—Ä–∞–ª–∏ (–ø—Ä–µ–¥—Å–∫–∞–∑–∞–ª–∏ False) –≤ –∏—Ç–æ–≥–µ: 1710 (35.00% –æ—Ç –¥–∞—Ç–∞—Å–µ—Ç–∞)
Recall(False) = 0.410
Recall(True)  = 0.948


#### –ò–Ω—Ç–µ—Ä–ø—Ä–µ—Ç–∞—Ü–∏—è —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤

    cutoff — фактически определяет, сколько объектов мы отбрасываем. Так как это фиксированные 40%, модель не выбирает порог по recall, а вы жёстко задаёте «хочу убрать ровно 40%».
    recall_true показывает, какую долю реальных True мы сохранили (среди оставшихся 60%). Если получается 0.95 или 0.90 — значит “потеряли” 5% или 10% True.
    recall_false –ø–æ–∫–∞–∑—ã–≤–∞–µ—Ç, –∫–∞–∫—É—é —á–∞—Å—Ç—å —Ä–µ–∞–ª—å–Ω—ã—Ö False –º—ã —É–±—Ä–∞–ª–∏ (–µ—Å–ª–∏ 0.80, –∑–Ω–∞—á–∏—Ç –º—ã –æ—Ç—Å–µ–∫–∞–µ–º 80% –≤—Å–µ—Ö False, –∞ 20% False –æ—Å—Ç–∞—é—Ç—Å—è).

Важно: этот метод управления работает по принципу “сначала выбираем долю (40%) для удаления, а уже потом смотрим, что там оказалось”. Если ваша реальная цель — убрать минимум 40% мусора и при этом терять не более 3% True, и выяснилось, что при отбрасывании ровно 40% теряется 10% True → модель не справляется с таким требованием.




- –û–±—â–µ–µ —á–∏—Å–ª–æ –æ–±—ä–µ–∫—Ç–æ–≤: 4886
- –£–¥–∞–ª–µ–Ω–æ (–ø—Ä–µ–¥—Å–∫–∞–∑–∞–Ω–æ –∫–∞–∫ False):—Ä–æ–≤–Ω–æ 35% –≤—ã–±–æ—Ä–∫–∏.
- Recall(False) = 0.410 означает, что среди всех действительно ложных сообществ вы «споймали» (и удалили) 41%.
- Recall(True) = 0.948 –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ —Å—Ä–µ–¥–∏ –≤—Å–µ—Ö –¥–µ–π—Å—Ç–≤–∏—Ç–µ–ª—å–Ω–æ —Ä–µ–ª–µ–≤–∞–Ω—Ç–Ω—ã—Ö —Å–æ–æ–±—â–µ—Å—Ç–≤ –≤—ã —Å–æ—Ö—Ä–∞–Ω–∏–ª–∏ 94.8% (—Ç–æ –µ—Å—Ç—å –ø–æ—Ç–µ—Ä—è–ª–∏ ~5.2% –∏—Å—Ç–∏–Ω–Ω–æ —Ä–µ–ª–µ–≤–∞–Ω—Ç–Ω—ã—Ö).

–ö–∞–∫ –∏–Ω—Ç–µ—Ä–ø—Ä–µ—Ç–∏—Ä–æ–≤–∞—Ç—å –Ω–∞ –ø—Ä–∞–∫—Ç–∏–∫–µ

- Вы “убираете” 35% датасета, которые выглядят наименее вероятными к классу True (релевантные).
- Из реального «мусора» (False) в результате отсеивается почти половина (41%).
- –ü—Ä–∏ —ç—Ç–æ–º –≤—ã —Å–ª—É—á–∞–π–Ω–æ –≤—ã–±—Ä–∞—Å—ã–≤–∞–µ—Ç–µ –æ–∫–æ–ª–æ 5.2% –Ω–∞—Å—Ç–æ—è—â–∏—Ö (True) —Å–æ–æ–±—â–µ—Å—Ç–≤.

Если вам подходит такой компромисс – убрать 40% мусора ценой 5% потерь релевантных сообществ, – то да, вы можете внедрить такой фильтр в сайт.